python/python/bystro/tada_mixture/mixture_tada_em.py

```diff
@@ -0,0 +1,394 @@
+could be naughty if the objective was multimodal, but I strongly doubt that
+this is the case.
+
+Objects
+-------
+MVTadaPoissonEM(K=4,training_options={})
+    This fits a standard poisson latent variable model.
+
+MVTadaZPoissonEM(K=4,training_options={})
+    This fits a standard ZIP latent variable model.
+
+Methods
+-------
+None
+
+"""
+import numpy as np
+from copy import deepcopy
+import scipy.stats as st # type: ignore
+from sklearn.mixture import GaussianMixture # type: ignore
+from tqdm import trange # type: ignore
+
+import torch
+from torch import nn
+from torch.nn import PoissonNLLLoss
+
+
+class MVTadaPoissonEM(object):
+    def __init__(self, K=4, training_options={}):
+        """
+        This is a Product Poisson latent variable model.
+
+        Parameters
+        ----------
+        K : int,default=4
+            The number of clusters, i.e. number of different types of 
+            genes
+
+        training_options : dict,default={}
+            The parameters for the inference scheme
+        """
+        self.K = int(K)
+        self.training_options = self._fill_training_options(training_options)
+
+    def fit(self, X, progress_bar=True, Lamb_init=None, pi_init=None):
+        """
+        Fits a model given count data X.
+
+        Parameters
+        ----------
+        X : np.array-like,shape=(N,p)
+            The count data
+
+        progress_bar : bool,default=True
+            Whether to print a progress bar while fitting
+
+        Lamb_init : np.array-like,shape=(k,p),default=None
+            Initialize the loadings of the model. Defaults to fitting a 
+            GMM on the data and using those.
+
+        pi_init : np.array-like,shape=(k),default=None
+            Initialize the class weights of the model.
+
+        Returns
+        -------
+        self
+        """
+        td = self.training_options
+        K = self.K
+
+        N, p = X.shape
+
+        # Initialize variables
+        if Lamb_init is None:
+            model = GaussianMixture(K)
+            model.fit(X)
+            Lambda_ = model.means_
+        else:
+            Lambda_ = Lamb_init
+
+        rng = np.random.default_rng(2021)
+        pi_ = rng.dirichlet(np.ones(K)) if pi_init is None else pi_init
+
+        myrange = trange if progress_bar else range
+
+        for i in myrange(td["n_iterations"]):
+            # E step compute most likely categories
+            Zprobs = np.zeros((N, K))
+            for j in range(K):
+                log_pmf = st.poisson.logpmf(X, Lambda_[j])
+                log_pmf_sum = np.sum(log_pmf, axis=1)
+                Zprobs[:, j] = log_pmf_sum + np.log(pi_[j])
+            Ez = np.argmax(Zprobs, axis=1)
+
+            # M step, maximize parameters
+            for j in range(K):
+                pi_[j] = np.mean(Ez == j)
+                Lambda_[j] = np.mean(X[Ez == j], axis=0)
+
+            eps = 0.001
+            pi_ = pi_ * (1 - K * eps) + np.ones(K) * eps
```

Comment:
it looks like this is a kind of "drag term" to regularize `pi_` and prevent it from ever assigning zero probability mass to any component?  might be worth a comment to that effect if so

Reply:
Yes, will do.